File: backend/app/schemas/calendar.py
import uuid
from datetime import date, time
from typing import Literal

from pydantic import BaseModel


class CalendarMilestoneItem(BaseModel):
    type: Literal["milestone"] = "milestone"
    id: uuid.UUID
    date: date
    name: str
    direction: str
    completed: bool
    actual_date: date | None
    project_id: uuid.UUID
    project_title: str


class CalendarAssignmentItem(BaseModel):
    type: Literal["assignment"] = "assignment"
    id: uuid.UUID
    date: date
    description: str
    status: str
    assignee_name: str
    milestone_id: uuid.UUID | None
    milestone_name: str | None
    project_id: uuid.UUID
    project_title: str


class CalendarItemsResponse(BaseModel):
    items: list[CalendarMilestoneItem | CalendarAssignmentItem]


class CalendarHolidayItem(BaseModel):
    type: Literal["holiday"] = "holiday"
    date: date
    name: str
    country: str


class CalendarHolidaysResponse(BaseModel):
    items: list[CalendarHolidayItem]


class CalendarEventItem(BaseModel):
    type: Literal["event"] = "event"
    event_id: uuid.UUID
    original_date: date
    date: date
    end_date: date
    title: str
    description: str | None
    all_day: bool
    start_time: time | None
    end_time: time | None
    about_user_name: str | None
    is_recurring: bool
    is_override: bool


class CalendarEventsResponse(BaseModel):
    items: list[CalendarEventItem]

File: frontend/src/api/calendar.ts
import { useQuery } from "@tanstack/react-query"

import { ApiError } from "@/api/auth"

export type CalendarMilestoneItem = {
  type: "milestone"
  id: string
  date: string
  name: string
  direction: string
  completed: boolean
  actual_date: string | null
  project_id: string
  project_title: string
}

export type CalendarAssignmentItem = {
  type: "assignment"
  id: string
  date: string
  description: string
  status: string
  assignee_name: string
  milestone_id: string | null
  milestone_name: string | null
  project_id: string
  project_title: string
}

export type CalendarItem = CalendarMilestoneItem | CalendarAssignmentItem
export type CalendarItemsResponse = { items: CalendarItem[] }

export type CalendarItemType = "milestone" | "assignment"

export type CalendarItemsParams = {
  start: string
  end: string
  department_id?: string[]
  client_id?: string[]
  discipline_id?: string[]
  types?: CalendarItemType[]
}

function qs(p: CalendarItemsParams): string {
  const u = new URLSearchParams()
  u.set("start", p.start)
  u.set("end", p.end)
  // Multi-select: emit one repeated param per id (FastAPI collects into a list).
  for (const id of p.department_id ?? []) u.append("department_id", id)
  for (const id of p.client_id ?? []) u.append("client_id", id)
  for (const id of p.discipline_id ?? []) u.append("discipline_id", id)
  if (p.types && p.types.length) u.set("types", p.types.join(","))
  return u.toString()
}

export type CalendarHolidayItem = {
  type: "holiday"
  date: string
  name: string
  country: string
}
export type CalendarHolidaysResponse = { items: CalendarHolidayItem[] }

export type CalendarHolidaysParams = { start: string; end: string }

export function useCalendarHolidays(params: CalendarHolidaysParams | null) {
  return useQuery<CalendarHolidaysResponse, ApiError>({
    queryKey: ["calendar", "holidays", params],
    queryFn: async () => {
      const u = new URLSearchParams({ start: params!.start, end: params!.end })
      const res = await fetch(`/api/calendar/holidays?${u.toString()}`, { credentials: "include" })
      const body = await res.json().catch(() => ({}))
      if (!res.ok) {
        const detail =
          body && typeof body === "object" && "detail" in body
            ? String((body as { detail: unknown }).detail)
            : "Load failed"
        throw new ApiError(detail, res.status)
      }
      return body as CalendarHolidaysResponse
    },
    enabled: !!params,
  })
}

export function useCalendarItems(params: CalendarItemsParams | null) {
  return useQuery<CalendarItemsResponse, ApiError>({
    queryKey: ["calendar", "items", params],
    queryFn: async () => {
      const res = await fetch(`/api/calendar/items?${qs(params!)}`, {
        credentials: "include",
      })
      const body = await res.json().catch(() => ({}))
      if (!res.ok) {
        const detail =
          body && typeof body === "object" && "detail" in body
            ? String((body as { detail: unknown }).detail)
            : "Load failed"
        throw new ApiError(detail, res.status)
      }
      return body as CalendarItemsResponse
    },
    enabled: !!params,
  })
}
